bsp_diff/common/kernel/linux-intel-lts2021/0032-enable-vblank-for-crtc.patch

```diff
@@ -0,0 +1,291 @@
+ void virtio_gpu_cursor_ack(struct virtqueue *vq);
++void virtio_gpu_vblank_ack(struct virtqueue *vq);
++void virtio_gpu_vblank_poll_arm(struct virtqueue *vq);
+ void virtio_gpu_fence_ack(struct virtqueue *vq);
+ void virtio_gpu_dequeue_ctrl_func(struct work_struct *work);
+ void virtio_gpu_dequeue_cursor_func(struct work_struct *work);
+ void virtio_gpu_dequeue_fence_func(struct work_struct *work);
+-
+ void virtio_gpu_notify(struct virtio_gpu_device *vgdev);
+-
++void virtio_gpu_vblankq_notify(struct virtio_gpu_device *vgdev);
+ int
+ virtio_gpu_cmd_resource_assign_uuid(struct virtio_gpu_device *vgdev,
+ 				    struct virtio_gpu_object_array *objs);
+diff --git a/drivers/gpu/drm/virtio/virtgpu_kms.c b/drivers/gpu/drm/virtio/virtgpu_kms.c
+index ad1a756ea151..670a7f112269 100644
+--- a/drivers/gpu/drm/virtio/virtgpu_kms.c
++++ b/drivers/gpu/drm/virtio/virtgpu_kms.c
+@@ -118,7 +118,7 @@ int virtio_gpu_find_vqs(struct virtio_gpu_device *vgdev)
+ 	struct virtqueue **vqs;
+ 	int i, total_vqs, err;
+ 	const char **names;
+-	int ret;
++	int ret = 0;
+ 
+ 	total_vqs = vgdev->num_vblankq + 2;
+ 	vqs = kcalloc(total_vqs, sizeof(*vqs), GFP_KERNEL);
+@@ -164,6 +164,7 @@ int virtio_gpu_init(struct virtio_device *vdev, struct drm_device *dev)
+ 	struct virtio_gpu_device *vgdev;
+ 	u32 num_scanouts, num_capsets;
+ 	int ret = 0;
++	int i;
+ 
+ 	if (!virtio_has_feature(vdev, VIRTIO_F_VERSION_1))
+ 		return -ENODEV;
+@@ -300,6 +301,12 @@ int virtio_gpu_init(struct virtio_device *vdev, struct drm_device *dev)
+ 
+ 	if (num_capsets)
+ 		virtio_gpu_get_capsets(vgdev, num_capsets);
++
++	virtio_gpu_vblankq_notify(vgdev);
++
++	for(i=0; i < vgdev->num_vblankq; i++)
++		virtqueue_disable_cb(vgdev->vblank[i].vblank.vq);
++
+ 	if (vgdev->has_edid)
+ 		virtio_gpu_cmd_get_edids(vgdev);
+ 	virtio_gpu_cmd_get_display_info(vgdev);
+diff --git a/drivers/gpu/drm/virtio/virtgpu_vq.c b/drivers/gpu/drm/virtio/virtgpu_vq.c
+index 4ee59ccbb456..af840566a3e2 100644
+--- a/drivers/gpu/drm/virtio/virtgpu_vq.c
++++ b/drivers/gpu/drm/virtio/virtgpu_vq.c
+@@ -31,8 +31,10 @@
+ #include <linux/virtio_config.h>
+ #include <linux/virtio_ring.h>
+ 
++#include <drm/drm.h>
+ #include <drm/drm_edid.h>
+ 
++#include <drm/drm_vblank.h>
+ #include "virtgpu_drv.h"
+ #include "virtgpu_trace.h"
+ 
+@@ -61,6 +63,60 @@ void virtio_gpu_ctrl_ack(struct virtqueue *vq)
+ 	schedule_work(&vgdev->ctrlq.dequeue_work);
+ }
+ 
++static void virtgpu_irqqueue_buf(struct virtqueue *vq,
++				  uint32_t *evtbuf)
++{
++	struct scatterlist sg[1];
++	sg_init_one(sg, evtbuf, sizeof(*evtbuf));
++	virtqueue_add_inbuf(vq, sg, 1, evtbuf, GFP_ATOMIC);
++}
++
++void virtio_gpu_vblank_poll_arm(struct virtqueue *vq)
++{
++	struct drm_device *dev = vq->vdev->priv;
++	struct virtio_gpu_device *vgdev = dev->dev_private;
++	unsigned long irqflags;
++	unsigned int len;
++	unsigned int *ret_value;
++	int target = 0;
++
++	while((target < vgdev->num_vblankq) && (vgdev->vblank[target].vblank.vq != vq)) {
++		target ++;
```

Comment:
no need for space between target and ++. Comment applies to the full PR changes

Reply:
OK